src/iop/rcd_demosaic.c

```diff
@@ -0,0 +1,645 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2010-2020 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+* RATIO CORRECTED DEMOSAICING
+* Luis Sanz Rodríguez (luis.sanz.rodriguez(at)gmail(dot)com)
+*
+* Release 2.3 @ 171125
+*
+* Original code from https://github.com/LuisSR/RCD-Demosaicing
+* Licensed under the GNU GPL version 3
+*/
+
+/*
+* The tiling and sse2 optimised coding has been done by
+* Ingo Weyrich (heckflosse67@gmx.de) in rawtherapee code and has been slightly
+* modified in cooperation.
+* Hanno Schwalm (hanno@schwalm-bremen.de)
+*/
+
+/* some notes about the algorithm
+  - the calculated data at the borders are not stable for tiling if we don't use
+    a RCD_BORDER of at least 9. The bordersize is the overlapping region for tiles.
+  - RCD_INTERPOLATE can be 6, we take the better data from rcd algorithm and only
+    border_interpolate the missing 4 and wrong 2 outermost rcd pixels for outer tiles.
+  - Is it safe to use -Ofast? Yes
+*/
+
+// this allows to pass RCD_TILESIZE to the code. On some machines larger tilesizes are faster
+// If it is undefined it will be set to 140, which is a good guess for modern x86/64 machines
+// Tested and tuned on Xeon E-2288G and AMD FX8350 
+#ifndef RCD_TILESIZE
+ #define RCD_TILESIZE 140
+#endif
+
+#ifdef __GNUC__
+  #pragma GCC optimize ("-Ofast")
+#endif
+
+#ifdef __GNUC__
+#define INLINE __inline
+#else
+#define INLINE inline
+#endif
+
+#ifdef __SSE__
+#include <x86intrin.h>
+#endif
+
+#define FCRCD(row, col) (cfarray[(((row) & 1)<<1) | ((col) & 1)])
+
+#define RCD_BORDER 9          // must be at least 9 to avoid tile-overlap errors
+#define RCD_INTERPOLATE 6     // we don't have to approximate the whole border
+#define RCD_TILEVALID (RCD_TILESIZE - 2 * RCD_BORDER)
+#define w1 RCD_TILESIZE
+#define w2 (2 * RCD_TILESIZE)
+#define w3 (3 * RCD_TILESIZE)
+#define w4 (4 * RCD_TILESIZE)
+
+#define eps 1e-5    // // Tolerance to avoid dividing by zero
+#define epssq 1e-10
+
+/* Some macros and inline functions taken from amaze_demosaic_RT */
+typedef __m128i vmask;
+typedef __m128 vfloat;
+
+#define F2V(a) _mm_set1_ps((a))
+#define LVFU(x) _mm_loadu_ps(&x)
+#define STVFU(x,y) _mm_storeu_ps(&x,y)
+#ifdef __SSE4_1__
```

Comment:
So we require SSE to be available (either SSE4 or SSE2) which is optional in current dt code. Are we sure we have SSE available in all the target dt runs on ?

Reply:
ok. i will add #if clause to enable the vector functions only in case we have see but there is only one internal loop that optionally takes sse code for speed.

Reply:
@TurboGit - dt can run on PowerPC which doesn't have sse (has altivec instead)... And it is possible that Apple M1 doesn't have sse and if by chance we'll have a way to support it then that's another platform without sse ;)

Reply:
Ok, maybe that's a good ground for playing with the clone targets ? This will enable the best hardware support and possible gain some more speed.

Reply:
Clone targets is gcc only, LLVM should have it but the feature is in limbo (or so I gather from discussion with @LebedevRI )

Reply:
Please, please don't add any new SSE-only code. (ideally, don't manually vectorize at all)
x86 will die one day, but it will sunset way before that, way before people currently expect.

Reply:
As @jenshannoschwalm wrote: The sse-code is optional. It just needs some more '#ifdef'

Reply:
Some more comments for all of us to decide on a more profound base.

The code is definitely **not** SSE-only.

It has been performance-tuned for sse enabled machines.

Unfortunately the above sse-tuned code (A) is not only about specific sse vector functions but also about the way we access the lp filter in rcd which affects how good the compiler can vectorize.

There is also a tested generic optimized version (B) fully trusting the compiler.

Measuring the performance shows for the whole rcd-demosaicing process
1. On SSE- enabled machines the sse optimized version (A) is ~6-8% faster than (B)
2. On non-SSE machines version (A) is probably ~25% slower than (B)

So there is a trade-off somehow, whether we should have the probably fastest code on current mainstream CPUs or something more generic. I prefer it the way i proposed in the pr but this not a very stron opinion. Let me know.

Reply:
@jenshannoschwalm 
> So there is a trade-off somehow, whether we should have the probably fastest code on current mainstream CPUs or something more generic.

There might be a possibility without a tradeoff: Use the >>1 indexing for the sse code and the normal indexing for non-SSE code. Just a thought...

Reply:
> Ok, maybe that's a good ground for playing with the clone targets ? This will enable the best hardware support and possible gain some more speed.

See my response on #7075, that clarifies the state of Clang/GCC/ICC multiversioning a bit.

Reply:
Would be a multiversioning for one out of 8 loops worth the effort to multiversion the whole algorithm? In my opinion SSE code for this one loop is much better to maintain...
The other loops are optimized pretty well by gcc, not as well by clang though...

Reply:
>This will enable the best hardware support and possible gain some more speed.

Fot the users which need best hardware (CPU) support nothing will beat a `-march=native` build.

Reply:
I don't really have the answer to that, typically we suggest that you benchmark to figure out when it is useful. That said, the multi-version function has a slight load-time performance hit (the 'resolution' of which version to use is handled when the library is loaded), and a slight call-time performance hit (since it is an indirect function call via a function pointer).

Depending on the performance gain possible for each version (particularly with SSE loops), this overhead is often more than acceptable.

Reply:
Just some more notes about new performance checks on my system E-2288G

1. gcc -march=native could not get an extra gain
2. so using  __DT_CLONE_TARGETS__ also didn't help
3. i checked using clang and the performance was worse, there were other dt pr's reporting this